apps/emqx/src/emqx_router_index.erl

```diff
@@ -0,0 +1,143 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2017-2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_router_index).
+
+-behaviour(gen_server).
+
+-include("emqx.hrl").
+-include("emqx_router.hrl").
+-include("types.hrl").
+-include("logger.hrl").
+
+-define(ROUTE_INDEX_TRIE, emqx_route_index_trie).
+
+-define(INDEXER, ?MODULE).
+
+-type update() :: {write | delete, emqx_types:topic(), emqx_router:dest()}.
+-type state() :: #{last_update := maybe(update())}.
+
+-export([
+    start_link/0,
+    enabled/0
+]).
+
+-export([match/1]).
+
+%% test only
+-export([peek_last_update/0]).
+
+%% gen_server callbacks
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+-spec start_link() -> startlink_ret().
+start_link() ->
+    case emqx_config:get([broker, perf, trie_local_async]) of
+        true ->
+            gen_server:start_link({local, ?INDEXER}, ?MODULE, [], []);
+        false ->
+            ignore
+    end.
+
+-spec enabled() -> boolean().
+enabled() ->
+    erlang:whereis(?INDEXER) =/= undefined.
+
+-spec match(emqx_types:topic()) -> [emqx_types:route()].
+match(Topic) ->
+    match_entries(Topic).
+
+-spec peek_last_update() -> maybe(update()).
+peek_last_update() ->
+    gen_server:call(?INDEXER, peek_last_update).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-spec init(_) -> {ok, state()}.
+init([]) ->
+    _ = emqx_trie:create_local(?ROUTE_INDEX_TRIE, [public, named_table]),
+    ok = mria:wait_for_tables([?ROUTE_TAB]),
+    % NOTE: it's unclear if we have guarantees that we won't miss any updates
+    {ok, _} = mnesia:subscribe({table, ?ROUTE_TAB, simple}),
+    ok = emqx_router:foldl_routes(fun(Route, _) -> insert_entry(Route) end, ok),
```

Comment:
Maybe move this to a post-init handle_info or handle_cast loopback message. (send it before `mnesia:subscribe`).
This is because the init/1 blocks supervisor, if the table has a lot of data, it may cause the boot sequence to slow down.

Reply:
We'll need to add some logging when at star/end of the index loop.
Also sync points for the boot sequence?
Ideally the MQTT listeners should not start until the index is completed.